Stop calling ranlib on created / installed libraries

shindere opened the pull request:
This PR is a follow-up to the discussion started in https://github.com/ocaml/ocaml/pull/11172#issuecomment-1092641466

It removes the calls to ranlib, both when libraries are installed and when they are created by ocamlmklib.

To make sure the timestamps of installed libraries match those of the object files they contain, we call the `install` program with the `-p` flag to preserve the timestamps when installing a file. This flag seems to be widely supported.

shindere commented:
I believe that the test failure observed on MacOS is not related to this PR.

xavierleroy commented on tools/ocamlmklib.ml:
```diff
@@ -25,10 +25,8 @@ let mklib out files opts =
-  else Printf.sprintf "%s rcs %s %s %s && %s %s"
-                      Config.ar out opts files Config.ranlib out
+  Printf.sprintf "link -lib -nologo %s-out:%s %s %s" machine out opts files
+  else Printf.sprintf "%s rcs %s %s %s" Config.ar out opts files
```
The `s` flag might be unnecessary and not available everywhere, according to https://pubs.opengroup.org/onlinepubs/9699919799/utilities/ar.html .  Removing the `s` flag would be consistent with the other invocations of `ar`.

xavierleroy reviewed (commented):
That's a nice simplification of our build system, thanks for taking care of this!

I started a round of CI precheck, let's see how this PR fares on strange systems.

shindere committed:
Stop calling ranlib on created / installed libraries

xavierleroy commented:
CI precheck seems happy, modulo the proverbial failures that seem unrelated.

I still suggest you remove the `s` flag from the `rcs` invocation of `ar`.

shindere commented:
Okay it seems comments by e-mail are broken.

It's the second time I try to comment on this PR and the second time it
fails.

Xavier Leroy (2022/04/14 06:56 -0700):
> CI precheck seems happy, modulo the proverbial failures that seem unrelated.
>
> I still suggest you remove the `s` flag from the `rcs` invocation of
> `ar`.

I did remove it and I did write a comment to this PR but God know where
it has gone.

I had two remarks / questions.

1. How abouot factorizing the code in `utils/ccomp.ml` and
`tools/ocamlmklib.ml` so that these hard-coded `ar` flags appear only
once in the source?

Although the Misc module is rather crowded already, would it be okay to
add a funciotn there?

2. In both these places and in the build system, special care is taken
to remove the library file before invoking `ar`. I assume this needs to
be kept? But then maybe we could factorize that, too (between the two
OCaml places, I mean).

xavierleroy commented:
> It's the second time I try to comment on this PR and the second time it fails.

Sorry for the email problems.

> How abouot factorizing the code in utils/ccomp.ml and
> tools/ocamlmklib.ml so that these hard-coded ar flags appear only
> once in the source?

`ar rc` is just a standard incantation, like `ls -l`.  I don't feel a strong need to share the `rc` part between the two uses.

Now there is the more general question of code duplication between `mklib` in tools/ocamlmklib.ml and `create_archive` in utils/ccomp.ml.  There is a bit of duplication, but there may be reasons why ocamlmklib.ml doesn't link with compiler-libs.

> In both these places and in the build system, special care is taken
> to remove the library file before invoking ar. I assume this needs to
> be kept?

I think so.  `ar rc` means "replace members in an archive file, creating it empty if it doesn't exist".  So, if you don't remove before calling `ar rc`, you could end up with old archive members left over.

> But then maybe we could factorize that, too (between the two OCaml places, I mean).

That's what I meant with the code duplication between `mklib` in tools/ocamlmklib.ml and `create_archive` in utils/ccomp.ml.   Not sure we want to spend more time on this, however.

xavierleroy reviewed (approved):
Looks very good to me, thanks!  Let's merge right now...

shindere commented:
Xavier Leroy (2022/04/14 10:17 -0700):
> `ar rc` is just a standard incantation, like `ls -l`.  I don't feel a strong need to share the `rc` part between the two uses.
>
> Now there is the more general question of code duplication between
> `mklib` in tools/ocamlmklib.ml and `create_archive` in utils/ccomp.ml.
> There is a bit of duplication, but there may be reasons why
> ocamlmklib.ml doesn't link with compiler-libs.

It indeed does not, but it links with \utils/misc` where we could add
the common code.

> > In both these places and in the build system, special care is taken
> > to remove the library file before invoking ar. I assume this needs to
> > be kept?
>
> I think so.  `ar rc` means "replace members in an archive file,
> creating it empty if it doesn't exist".  So, if you don't remove
> before calling `ar rc`, you could end up with old archive members left
> over.

Yes, makes sense, indeed.

> > But then maybe we could factorize that, too (between the two OCaml places, I mean).
>
> That's what I meant with the code duplication between `mklib` in
> tools/ocamlmklib.ml and `create_archive` in utils/ccomp.ml.   Not sure
> we want to spend more time on this, however.

Ultimately, anothe roption, at least for the build system, would be to
figure out how much we could rely on `libtool` for these kind of things,
since we do already embed it.

But that would of course not help in any way for the compiler itself. I
was still wondering whether it wouldn't be better to just call the C
compiler, which would perhaps take care of the details anyway. I think
we did that already to simplify the invocation of the assembler on some
platforms.

xavierleroy commented:
Cherry-picked to 4.14: 757e0a718 .

monniaux commented:
To whom it may be of interest: calling `ranlib` on multiple files fails on OpenBSD (use of `llvm-ranlib`) too and prevents installing ocaml on that system.

shindere commented:
Xavier Leroy (2022/04/13 10:37 -0700):
> @xavierleroy commented on this pull request.
>
> That's a nice simplification of our build system, thanks for taking
> care of this!

And thanks for the positive feedback.

> > @@ -25,10 +25,8 @@ let mklib out files opts =
>      then "-machine:AMD64 "
>      else ""
>    in
> -  Printf.sprintf "link -lib -nologo %s-out:%s %s %s"
> -                 machine out opts files
> -  else Printf.sprintf "%s rcs %s %s %s && %s %s"
> -                      Config.ar out opts files Config.ranlib out
> +  Printf.sprintf "link -lib -nologo %s-out:%s %s %s" machine out opts files
> +  else Printf.sprintf "%s rcs %s %s %s" Config.ar out opts files
>
> The `s` flag might be unnecessary and not available everywhere,
> according to
> https://pubs.opengroup.org/onlinepubs/9699919799/utilities/ar.html .
> Removing the `s` flag would be consistent with the other invocations
> of `ar`.

I removed it. Thanks. How about factorizing at least the function that
takes the names of the object files and libraries and returns the
command in, say, the Misc module?
Although it is now fixed, the discrepancy between the way `ar` is called
from different places feels like a code smell to me.

Also: at the moment we make sure to remove the library before invoking
ar, both in the OCaml code and in our build system.

Is such a precaution still useful? I wold tend to think it's not, since
`r` does replace members even if they are already present in the
archive.

shindere commented:
Xavier Leroy (2022/04/14 06:56 -0700):
> CI precheck seems happy, modulo the proverbial failures that seem unrelated.
>
> I still suggest you remove the `s` flag from the `rcs` invocation of
> `ar`.

I did remove it and I did write a comment to this PR but God know where
it has gone.

I had two remarks / questions.

1. How abouot factorizing the code in `utils/ccomp.ml` and
`tools/ocamlmklib.ml` so that these hard-coded `ar` flags appear only
once in the source?

Although the Misc module is rather crowded already, would it be okay to
add a funciotn there?

2. In both these places and in the build system, special care is taken
to remove the library file before invoking `ar`. I assume this needs to
be kept? But then maybe we could factorize that, too (between the two
OCaml places, I mean).

kit-ty-kate commented:
just for information, this change triggered a failure in `core_unix` on macOS, where it was previously just a warning (see the fix https://github.com/janestreet/core_unix/pull/5)

damiendoligez commented:
xref #1494

xavierleroy commented:
Ah, macOS and its famously finicky development tools...

If we were to use `ar rcs` instead of `ar rc`, would this improve the situation?  I.e. force `ar` to create a symbol table even if the archive is empty, making `ld` happy about the archive.

dra27 commented:
For 4.14.1, should we revert this PR and fix the problem more exactly just by calling ranlib with one file argument in the build system?
